File: data/src/main/java/com/balv/imdb/data/repository/AppRepo.java
package com.balv.imdb.data.repository;

import android.annotation.SuppressLint;
import android.text.TextUtils;
import android.util.Log;

import androidx.paging.Pager;
import androidx.paging.PagingConfig;
import androidx.paging.PagingData;
import androidx.paging.PagingDataTransforms;
import androidx.paging.rxjava2.PagingRx;

import com.balv.imdb.data.Constant;
import com.balv.imdb.data.executor.PagingExecutor;
import com.balv.imdb.data.local.AppDb;
import com.balv.imdb.data.local.UserPreference;
import com.balv.imdb.data.mapper.Mapper;
import com.balv.imdb.data.mediator.PageKeyedRemoteMediator;
import com.balv.imdb.data.model.MovieEntity;
import com.balv.imdb.data.network.ApiService;
import com.balv.imdb.domain.models.ApiResult;
import com.balv.imdb.domain.models.ErrorResult;
import com.balv.imdb.domain.models.Movie;
import com.balv.imdb.domain.repositories.IMovieRepository;

import java.util.List;
import java.util.concurrent.Executor;
import java.util.stream.Collectors;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import retrofit2.HttpException;

@SuppressLint("CheckResult")
public class AppRepo implements IMovieRepository {
    private static final String TAG = "AppRepo";

    public AppDb mAppDb;

    public ApiService mApiService;

    private static final int CPU_COUNT = Runtime.getRuntime().availableProcessors();
    private static final int CORE_POOL_SIZE = Math.max(2, Math.min(CPU_COUNT - 1, 4));
    private static final int MAXIMUM_POOL_SIZE = CPU_COUNT * 2;

    @Inject
    Mapper mapper;

    @Inject
    UserPreference mUserPref;

    @Inject
    public AppRepo(AppDb mAppDb, ApiService mApiService, Mapper nwToDbMapper, UserPreference preference) {
        this.mapper = nwToDbMapper;
        this.mAppDb = mAppDb;
        this.mApiService = mApiService;
        this.mUserPref = preference;
    }

    @Override
    public Observable<List<Movie>> getLocalMovieData() {
        return mAppDb.movieDao().getMainMovieList().map(list ->
                list.stream()
                        .map(movieEntity -> mapper.entityToDomain(movieEntity))
                        .collect(Collectors.toList()));

    }

    @Override
    public Observable<Movie> getMovieDetail(String id) {
        return mAppDb.movieDao().setMovieDetailSource(id)
                .map(mapper::entityToDomain);
    }

    public static final Executor EXECUTOR = new PagingExecutor (CORE_POOL_SIZE, MAXIMUM_POOL_SIZE);
    @Override
    public Observable<ApiResult> getDetailFromNetwork(String id) {
        return mApiService.getMovieDetail(Constant.API_KEY, id)
                .subscribeOn(Schedulers.io())
                .map(result -> {
                    mAppDb.movieDao().updateMovies(mapper.detailToEntity(result));
                    return new ApiResult(result);
                });
    }

    @Override
    public Observable<ApiResult> getNextRemoteDataPage(int page) {
        return mApiService.getMoviesList(Constant.API_KEY, Constant.MOVIE_KEYWORD, page)
                .subscribeOn(Schedulers.io())
                .map(result -> {
                    Log.i(TAG, "getRemoteData: listSize=" + result.list.size());
                    List<MovieEntity> entityList = result.list.stream()
                            .map(networkObject -> mapper.networkToEntity(networkObject))
                            .collect(Collectors.toList());
                    mAppDb.movieDao().insertAll(entityList);

                    return new ApiResult(result);
                })
                .onErrorResumeNext(e -> {
                    Log.i(TAG, "getNextRemoteDataPage: error" + e);
                    if (e instanceof HttpException && ((HttpException) e).code() == Constant.HTTP_ERROR_SESSION_EXPIRED)
                        return Observable.just(new ApiResult(new ErrorResult(401, e.getMessage())));
                    else
                        return Observable.just(new ApiResult(new ErrorResult(Constant.NETWORK_ERROR, e.getMessage())));
                });
    }


    @Override
    public int getCurrentDataSize() {
        return mAppDb.movieDao().getCount();
    }

    @Override
    public long getDataRefreshDate() {
        return mUserPref.getLongValue(Constant.DATA_REFRESH);
    }

    @Override
    public void setDataRefreshDate(long time) {
        mUserPref.saveLongValue(Constant.DATA_REFRESH, time);
    }

    @Override
    public Observable<PagingData<Movie>> getPagingData(  ) {
        PagingConfig config = new PagingConfig(10, 10,
                true, 10, 200);
        Pager<Integer, MovieEntity> pager = new Pager(config, 1,
                new PageKeyedRemoteMediator(this, mAppDb, mApiService, mapper),
                () -> mAppDb.movieDao().pagingSource());

        Observable<PagingData<Movie>> pagingDataObservable = PagingRx.getObservable(pager)
                .map(paging -> PagingDataTransforms.map(paging, EXECUTOR, mapper::entityToDomain));
        return pagingDataObservable;

    }


}

File: data/src/main/java/com/balv/imdb/data/repository/AppRepo.kt
package com.balv.imdb.data.repository

import android.annotation.SuppressLint
import android.util.Log
import androidx.paging.ExperimentalPagingApi
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.map
import com.balv.imdb.data.Constant
import com.balv.imdb.data.local.AppDb
import com.balv.imdb.data.local.UserPreference
import com.balv.imdb.data.mapper.Mapper
import com.balv.imdb.data.mediator.PageKeyedRemoteMediator
import com.balv.imdb.data.model.SearchData
import com.balv.imdb.data.network.ApiService
import com.balv.imdb.domain.models.ApiResult
import com.balv.imdb.domain.models.ErrorResult
import com.balv.imdb.domain.models.Movie
import com.balv.imdb.domain.repositories.IMovieRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import retrofit2.HttpException
import javax.inject.Inject

@SuppressLint("CheckResult")
class AppRepo @Inject constructor(
    private val appDb: AppDb,
    private val apiService: ApiService,
    private val userPref: UserPreference
) : IMovieRepository {

    override suspend fun getMovieDetailLocal(id: Int): Flow<Movie?> {
        return appDb.movieDao().getMovieDetailLocal(id)
            .map { input-> input?.let { Mapper.entityToDomain(it) } }
    }

    override suspend fun getDetailFromNetwork(id: Int): ApiResult<Movie?> {
        return getRemoteResponse {
            apiService.getMovieDetail(movieId = id).let { result->
                val localEntity = Mapper.remoteMovieToEntity(result).copy(
                    polledDate = System.currentTimeMillis()
                )
                appDb.movieDao().updateMovies(localEntity)
                Mapper.networkToDomain(result)
            }
        }
    }

    override suspend fun getNextRemoteDataPage(page: Int): ApiResult<SearchData> {
        return getRemoteResponse {
            apiService.discoverMovies(page = page)
                .let { result: SearchData ->
                    Log.i(TAG, "getRemoteData: listSize=" + result.results.size)
                    result.results.map { remote ->
                        Mapper.remoteMovieToEntity(remote)
                    }.also {
                        appDb.movieDao().insertAll(it)
                    }
                    result
                }
        }
    }

    private suspend fun <T> getRemoteResponse(block: suspend () -> T): ApiResult<T> {
        return kotlin.runCatching {
            ApiResult.success(block())
        }.getOrElse { e ->
            if (e is HttpException && e.code() == Constant.HTTP_ERROR_SESSION_EXPIRED) {
                ApiResult.error(ErrorResult(401, e.message.orEmpty()))
            } else {
                ApiResult.error(ErrorResult(Constant.NETWORK_ERROR, e.message.orEmpty()))
            }
        }
    }


    override suspend fun getDataRefreshDate(): Long {
        return userPref.getLongValue(Constant.DATA_REFRESH)
    }

    override suspend fun setDataRefreshDate(date: Long) {
        userPref.saveLongValue(Constant.DATA_REFRESH, date)
    }

    @OptIn(ExperimentalPagingApi::class)
    override fun allMoviesPaging(): Flow<PagingData<Movie>> {
        val config = PagingConfig(
            10, 1,
            true, 10, 30
        )
        val pager = Pager(
            config, 1,
            PageKeyedRemoteMediator(this, appDb, apiService)
        ) { appDb.movieDao().pagingSource() }

        return pager.flow
            .map { pagingData ->
                pagingData.map { input -> input.let { Mapper.entityToDomain(it) } }
            }
    }

    override fun getPopularMoviesFetchedDate(): Long =
        userPref.getLongValue(PREF_POPULAR_FETCHED_DATE)

    override fun getPopularMovies(): Flow<List<Movie>> {
        return appDb.movieDao().getTop20PopularMovies().map {
            Log.i(TAG, "getPopularMovies: $it")
            it.map { movieEntity -> Mapper.entityToDomain(movieEntity) }
        }
    }

    override suspend fun refreshPopularMovies() {
        getRemoteResponse {
            apiService.getPopularMovies()
        }.data?.results?.let { listMv ->
            userPref.saveLongValue(PREF_POPULAR_FETCHED_DATE, System.currentTimeMillis())
            appDb.movieDao().insertAll(
                listMv.map {
                    Mapper.remoteMovieToEntity(it)
                }
            )
        }
    }

    companion object {
        private const val TAG = "AppRepo"
        private const val PREF_POPULAR_FETCHED_DATE = "PREF_POPULAR_FETCHED_DATE"
    }
}
